core/api.go

```diff
@@ -344,3 +341,46 @@ func (a *api) getFileConfig(g *gin.Context, mill m.Mill, use string, plaintext b
 func (a *api) abort500(g *gin.Context, err error) {
 	g.String(http.StatusInternalServerError, err.Error())
 }
+
+// getCORSSettings returns custom CORS settings given HTTPHeaders options
+func getCORSSettings(config *config.Config) cors.Config {
+	headers := config.API.HTTPHeaders
+	cconfig := cors.DefaultConfig()
+
+	control, ok := headers["Access-Control-Allow-Origin"]
+	if ok && len(control) > 0 {
+		cconfig.AllowOrigins = control
+		for _, origin := range control {
+			if origin == "*" {
+				cconfig.AllowAllOrigins = true
+				cconfig.AllowOrigins = nil
+			}
+		}
+	} else {
+		// By default use API origin host
+		defaultHost := config.Addresses.API
+		match, _ := regexp.MatchString("^https?://", defaultHost)
```

Comment:
This feels messy to me, but the cors package requires http{s} prefixes on origins... ipfs for example, hard-codes:
localhosts: https://github.com/ipfs/go-ipfs/blob/af73c5097ce0ff70a18db25f801b34fa66a98202/core/corehttp/commands.go#L40